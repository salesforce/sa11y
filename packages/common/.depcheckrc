# Ignore webdriverio - we need the types for `WebdriverIO.Browser`
ignores: ['webdriverio']
