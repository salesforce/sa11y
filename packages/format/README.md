# `format`

Format accessibility results from axe

## Usage

```
import axe from 'axe-core';
import { a11yResultsFormatter } from '@sa11y/format';

const results = await axe.run();
console.log(a11yResultsFormatter(results.violations));
```
