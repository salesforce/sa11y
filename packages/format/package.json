{
    "name": "@sa11y/format",
    "version": "0.2.1-beta.0",
    "description": "Accessibility results re-formatter",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/format#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/format"
    },
    "keywords": [
        "accessibility",
        "automated testing",
        "axe",
        "results",
        "format",
        "template"
    ],
    "main": "dist/format.js",
    "types": "dist/format.d.ts",
    "files": [
        "dist/**/*.js"
    ],
    "dependencies": {
        "axe-core": "3.5.5"
    },
    "devDependencies": {
        "@sa11y/test-utils": "0.2.0-beta.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
