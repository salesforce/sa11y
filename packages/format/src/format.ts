/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Result } from 'axe-core';

/**
 * Formatter defines the function signature to format accessibility violations found by axe
 */
export interface Formatter {
    (violations: Result[]): string;
}

/**
 * Format accessibility results from axe
 * @param violations - Result list from axe
 * */
// TODO: Add handlebars template for formatting
// TODO: Add support for different output formats console(colored), plain text, HTML, xUnit
export function a11yResultsFormatter(violations: Result[]): string {
    return violations
        .map((violation) => {
            return violation.nodes
                .map((node) => {
                    const selectors = node.target.join(', ');
                    return `${violation.help} (${violation.id}): ${selectors}\n  - More info: ${violation.helpUrl}`;
                })
                .join('\n\n');
        })
        .join('\n\n');
}
