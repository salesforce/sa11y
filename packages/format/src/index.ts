/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yError, Options } from './format';
export { exceptionListFilter } from './filter';
export { A11yResult, A11yResults } from './result';
