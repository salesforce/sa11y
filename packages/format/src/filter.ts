/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as axe from 'axe-core';
import { AxeResults, ExceptionList } from '@sa11y/common';

/**
 * Filter a11y violations from axe based on given {@link ExceptionList}
 * @param violations - List of violations found with axe
 * @param exceptionList - {@link ExceptionList} of map of rule to corresponding css targets that needs to be filtered from a11y results
 */
export function exceptionListFilter(violations: AxeResults, exceptionList: ExceptionList = {}): AxeResults {
    const exceptionRules = Object.keys(exceptionList);
    if (exceptionRules.length === 0) return violations;

    const filteredViolations: AxeResults = [];

    for (const violation of violations) {
        if (!exceptionRules.includes(violation.id)) {
            filteredViolations.push(violation);
        } else {
            for (const result of violation.nodes) {
                const filteredResults = [];
                result.target.forEach((cssSelectorItem) => {
                    if (Array.isArray(cssSelectorItem)) {
                        cssSelectorItem.forEach((cssSelector) => {
                            if (!exceptionList[violation.id].includes(cssSelector)) {
                                filteredResults.push(cssSelector);
                            }
                        });
                    } else {
                        if (!exceptionList[violation.id].includes(cssSelectorItem)) {
                            filteredResults.push(cssSelectorItem);
                        }
                    }
                });
                if (filteredResults.length > 0) {
                    filteredViolations.push(violation);
                }
            }
        }
    }

    return filteredViolations;
}

/**
 * Filter a11y violations from axe based on given selectors filter keywords
 * @param violations - List of violations found with axe
 * @param selectorFilterKeywords - List of selector keywords to filter violations for
 */
export function exceptionListFilterSelectorKeywords(
    violations: AxeResults,
    selectorFilterKeywords: string[]
): AxeResults {
    const filteredViolations: AxeResults = [];
    for (const violation of violations) {
        const filteredNodes: axe.NodeResult[] = [];
        for (const node of violation.nodes) {
            const isSelectorFilterKeywordsExists = checkSelectorFilterKeyWordsExists(node, selectorFilterKeywords);
            if (!isSelectorFilterKeywordsExists) {
                filteredNodes.push(node);
            }
        }
        if (filteredNodes.length > 0) {
            violation.nodes = filteredNodes;
            filteredViolations.push(violation);
        }
    }
    return filteredViolations;
}

function checkSelectorFilterKeyWordsExists(node: axe.NodeResult, selectorFilterKeywords: string[]): boolean {
    const selectorAncestry = (node.ancestry?.flat(Infinity) ?? []) as string[];
    let isExists = false;
    selectorFilterKeywords.some((keyword) => {
        isExists = selectorAncestry.some((selector) => {
            const lastSelector = selector.split('>').pop();
            return lastSelector?.includes(keyword) ?? false;
        });
        return isExists;
    });
    return isExists;
}
