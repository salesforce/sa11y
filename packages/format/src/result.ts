/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from '@sa11y/common';
import { NodeResult, Result } from 'axe-core';
import { priorities, wcagLevels, WcagMetadata } from '@sa11y/preset-rules';

const defaultImpact = 'minor'; // if impact is undefined
const formatSpacing = '\t'.repeat(12);
// Helper object to sort violations by impact order
const impactOrder = {
    critical: 1,
    serious: 2,
    moderate: 3,
    minor: 4,
};

export class A11yResults {
    private static consolidated = new Map<string, string[]>();

    /**
     * Clear accumulated consolidated results
     */
    static clear(): void {
        this.consolidated.clear();
    }

    /**
     * Consolidate given a11y results based on the given key (test scope)
     * and return new results that are not already present
     */
    static add(results: A11yResult[], key = ''): A11yResult[] {
        const existingResults = this.consolidated.get(key) || [];
        if (existingResults.length === 0) this.consolidated.set(key, existingResults);
        return results.filter((result) => {
            if (!existingResults.includes(result.key)) {
                existingResults.push(result.key);
                return result;
            }
        });
    }

    /**
     * Sort a11y violations from axe in order of impact
     */
    static sort(violations: AxeResults): AxeResults {
        return violations.sort((a, b) => {
            const aImpact = impactOrder[a.impact || defaultImpact];
            const bImpact = impactOrder[b.impact || defaultImpact];
            if (aImpact < bImpact) return -1;
            if (aImpact > bImpact) return 1;
            return 0;
        });
    }

    /**
     * Normalize and flatten a11y violations from Axe
     */
    static convert(violations: AxeResults): A11yResult[] {
        return A11yResults.sort(violations).flatMap((violation) => {
            return violation.nodes.map((node) => {
                return new A11yResult(violation, node);
            });
        });
    }
}

/**
 * Filtered a11y result containing selected and normalized info about the a11y failure
 */
export class A11yResult {
    // Note: This is serialized as part of A11yError and read back from the
    // custom test results processor to consolidate/transform a11y errors.
    // So it can only have members that can be deserialized back easily
    // i.e. no object refs etc.
    public readonly id: string;
    public readonly selectors: string;
    public readonly html: string;
    public readonly description: string;
    public readonly helpUrl: string;
    public readonly wcag: string;
    public readonly summary: string;
    public readonly key: string; // Represent a key with uniquely identifiable info
    public readonly ancestry: string;
    public readonly any: string;
    public readonly all: string;
    public readonly none: string;
    private readonly wcagData: WcagMetadata; // Used to sort results

    constructor(violation: Result, node: NodeResult) {
        this.id = violation.id;
        this.description = violation.help;
        this.wcagData = new WcagMetadata(violation);
        this.wcag = this.wcagData.toString();
        this.helpUrl = violation.helpUrl.split('?')[0];
        this.selectors = node.target.sort().join('; ');
        this.html = node.html;
        // TODO (code cov): Add test data where failure summary is missing
        /* istanbul ignore next */
        this.summary = node.failureSummary || '';
        this.key = `${this.id}--${this.selectors}`;
        this.ancestry = (node.ancestry?.flat(Infinity) ?? []).join('\n');
        this.any = node.any?.map((item) => `${formatSpacing}• ${item.message}`).join('\n');
        this.all = node.all?.map((item) => `${formatSpacing}• ${item.message}`).join('\n');
        this.none = node.none?.map((item) => `${formatSpacing}• ${item.message}`).join('\n');
    }

    /**
     * Sort results by Priority and WCAG Level
     */
    static sort(results: A11yResult[]): A11yResult[] {
        return results.sort((a, b) => {
            const priorityA = priorities.indexOf(a.wcagData.priority);
            const priorityB = priorities.indexOf(b.wcagData.priority);
            const wcagLevelA = wcagLevels.indexOf(a.wcagData.wcagLevel);
            const wcagLevelB = wcagLevels.indexOf(b.wcagData.wcagLevel);
            return priorityA - priorityB || wcagLevelA - wcagLevelB;
        });
    }
}

/**
 * Adds WCAG Success Criteria Info to result violations
 * @param results - Array of Axe results
 */
export function appendWcag(results: AxeResults) {
    for (const violation of results) {
        const wcagMetadata = new WcagMetadata(violation);
        violation.tags.push(wcagMetadata.toString());
    }
}
