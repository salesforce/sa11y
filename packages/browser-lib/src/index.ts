/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { exceptionListFilter, appendWcag } from '@sa11y/format';
import { defaultRuleset, registerCustomRules } from '@sa11y/preset-rules';
export { base, extended, full } from '@sa11y/preset-rules';
export const namespace = 'sa11y';

/**
 * Wrapper function to check accessibility to be invoked after the sa11y minified JS file
 * is injected into the browser
 * @param scope - Scope of the analysis as {@link A11yCheckableContext}
 * @param rules - preset sa11y rules defaulting to {@link base}
 * @param exceptionList - mapping of rule to css selectors to be filtered out using {@link exceptionListFilter}
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function checkAccessibility(
    scope = document,
    rules = defaultRuleset,
    exceptionList = {},
    addWcagInfo = false
) {
    // TODO (debug): adding type annotations to args, return type results in error:
    //  "[!] Error: Unexpected token" in both rollup-plugin-typescript2 and @rollup/plugin-typescript
    //  Compiling the index.ts file with tsc and using the dist/index.js file results
    //  in error when importing the "namespace" var. This would probably be easier to fix
    //  which could then result in getting rid of the rollup typescript plugins.

    // To register custom rules
    registerCustomRules();

    const results = await axe.run(scope || document, rules);
    const filteredResults = exceptionListFilter(results.violations, exceptionList);

    if (addWcagInfo) {
        appendWcag(filteredResults);
    }
    return JSON.stringify(filteredResults);
}
