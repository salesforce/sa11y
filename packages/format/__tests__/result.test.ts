/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { A11yResult, A11yResults, appendWcag } from '../src';
import { getViolations } from './format.test';
import { AxeResults } from '@sa11y/common';
import { NodeResult, Result } from 'axe-core';
import { expect } from '@jest/globals';

const a11yIssues = [
    { impact: undefined },
    { impact: undefined },
    { impact: 'moderate' },
    { impact: 'minor' },
    { impact: 'critical' },
    { impact: 'critical' },
] as AxeResults;

let violations: AxeResults = [];
let a11yResults: A11yResult[] = [];
beforeAll(async () => {
    violations = await getViolations();
    a11yResults = A11yResults.convert(violations);
});
beforeEach(() => A11yResults.clear());

describe('a11y result', () => {
    it('should be serializable', () => {
        const deserializeResults = (a11yResults) => JSON.parse(JSON.stringify(a11yResults)) as A11yResult[];
        const a11yResults = A11yResults.convert(violations);
        expect(deserializeResults(a11yResults)).toEqual(a11yResults);

        // Add a non-serializable property to A11yResult (E.g. a function)
        a11yResults.map(
            (a11yResult) =>
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore TS2339: Property 'foo' does not exist on type 'A11yResult'.
                (a11yResult.foo = () => {
                    return 'foo';
                })
        );
        expect(deserializeResults(a11yResults)).not.toEqual(a11yResults);
    });

    it('should sort a11y issues by impact', () => {
        A11yResults.sort(a11yIssues);
        expect(a11yIssues[0].impact).toBe('critical');
        expect(a11yIssues[1].impact).toBe('critical');
        expect(a11yIssues[2].impact).toBe('moderate');
        expect(a11yIssues[3].impact).toBeUndefined(); // Sort by "defaultImpact"
        expect(a11yIssues[4].impact).toBeUndefined();
        expect(a11yIssues[5].impact).toBe('minor');
    });

    it('should sort a11y results by priority, WCAG level', () => {
        // Create test data with rule ids of different priority
        const rules = [
            'identical-links-same-purpose', // P3, AAA
            'avoid-inline-spacing', // P3, AA
            'autocomplete-valid', // P2, AA
            'audio-caption', // P1, A
        ];
        const expectedSortedRules = [...rules].reverse();
        const a11yResults = A11yResults.convert(
            rules.map(
                (rule) =>
                    ({
                        id: rule,
                        helpUrl: 'foo',
                        nodes: [{ target: ['foo'], html: 'bar', failureSummary: 'baz' } as NodeResult],
                    } as Result)
            )
        );

        // baseline check before sorting
        a11yResults.forEach((result, index) => expect(result.id).toEqual(rules[index]));

        // Sort and check if they have been rearranged by priority and WCAG level
        A11yResult.sort(a11yResults).forEach((result, index) => expect(result.id).toEqual(expectedSortedRules[index]));
    });

    it('should consolidate violations', () => {
        expect(a11yResults.length).toBeGreaterThan(0);
        expect(A11yResults.add(a11yResults)).toHaveLength(a11yResults.length);
    });

    it('should not add the same violations again', () => {
        expect(A11yResults.add(a11yResults)).toHaveLength(a11yResults.length);
        expect(A11yResults.add(a11yResults)).toHaveLength(0);
        expect(A11yResults.add(a11yResults.concat(a11yResults))).toHaveLength(0);
    });

    it('should not add a single duplicate violation', () => {
        expect(A11yResults.add(a11yResults)).toHaveLength(a11yResults.length);
        const a11yResult = a11yResults[0];
        // Shouldn't add an individual duplicate violation
        expect(A11yResults.add([a11yResult])).toHaveLength(0);

        // Shouldn't add modified violations (after a result is removed)
        expect(A11yResults.add(a11yResults)).toHaveLength(0);

        // Should add the result copy with diff id
        const violation = violations[0];
        // base line check before modifying id, css
        const existingA11yResult = new A11yResult(violation, violation.nodes[0]);
        expect(A11yResults.add([existingA11yResult])).toHaveLength(0);

        // Create a copy with diff ID
        const newA11yResultId = new A11yResult({ ...violation, id: 'foo' }, violation.nodes[0]);
        expect(A11yResults.add([newA11yResultId])).toHaveLength(1);
        expect(A11yResults.add([newA11yResultId])).toHaveLength(0);

        // Should add the result copy with diff CSS selector
        const newA11yResultCss = new A11yResult(violation, { ...violation.nodes[0], target: ['bar'] });
        expect(A11yResults.add([newA11yResultCss])).toHaveLength(1);
        expect(A11yResults.add([newA11yResultCss])).toHaveLength(0);
    });

    it.each(['', 'foo', 'bar'])('should consolidate based on given key: %#', (key) => {
        expect(A11yResults.add(a11yResults, key)).toHaveLength(a11yResults.length);
        expect(A11yResults.add(a11yResults, key)).toHaveLength(0);
    });
});

describe('appendWcag', () => {
    it('should add WCAG metadata to each result', () => {
        appendWcag(violations);

        // Expect that WCAG metadata is added to each result
        expect(violations[0].tags).toContain('SA11Y-WCAG-SC2.4.2-P2');
        expect(violations[1].tags).toContain('SA11Y-WCAG-SC3.1.1-P2');
    });

    it('should handle results with no WCAG metadata', () => {
        appendWcag(violations);

        // Expect that no additional tags are added to results with no WCAG metadata
        expect(violations[0].tags).not.toContain('SA11Y-WCAG');
        expect(violations[1].tags).not.toContain('SA11Y-WCAG');
    });
});
