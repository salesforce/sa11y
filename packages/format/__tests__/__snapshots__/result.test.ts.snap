// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 2`] = `2`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 3`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 1`] = `""`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 2`] = `0`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 3`] = `
"0 Accessibility issues found
 "
`;

exports[`a11y Results Formatter should format using specified options: 0 1`] = `"[{\\"id\\":\\"document-title\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.text-alternatives\\",\\"wcag2a\\",\\"wcag242\\",\\"TTv5\\",\\"TT12.a\\",\\"EN-301-549\\",\\"EN-9.2.4.2\\",\\"ACT\\"],\\"description\\":\\"Ensures each HTML document contains a non-empty <title> element\\",\\"help\\":\\"Documents must have <title> element to aid in navigation\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.8/document-title?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"doc-has-title\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Document does not have a non-empty <title> element\\"}],\\"all\\":[],\\"none\\":[],\\"impact\\":\\"serious\\",\\"html\\":\\"<html lang=\\\\\\"en\\\\\\"><head><style>* { pointer-events: all }</style></head><body>\\\\n\\\\n<div id=\\\\\\"dom-with-issues\\\\\\">\\\\n    <a href=\\\\\\"#\\\\\\"></a>\\\\n</div>\\\\n\\\\n\\\\n</body></html>\\",\\"target\\":[\\"html\\"],\\"failureSummary\\":\\"Fix any of the following:\\\\n  Document does not have a non-empty <title> element\\"}]},{\\"id\\":\\"link-name\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.name-role-value\\",\\"wcag2a\\",\\"wcag244\\",\\"wcag412\\",\\"section508\\",\\"section508.22.a\\",\\"TTv5\\",\\"TT6.a\\",\\"EN-301-549\\",\\"EN-9.2.4.4\\",\\"EN-9.4.1.2\\",\\"ACT\\"],\\"description\\":\\"Ensures links have discernible text\\",\\"help\\":\\"Links must have discernible text\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.8/link-name?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"has-visible-text\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element does not have text that is visible to screen readers\\"},{\\"id\\":\\"aria-label\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-label attribute does not exist or is empty\\"},{\\"id\\":\\"aria-labelledby\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\"},{\\"id\\":\\"non-empty-title\\",\\"data\\":{\\"messageKey\\":\\"noAttr\\"},\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element has no title attribute\\"}],\\"all\\":[],\\"none\\":[{\\"id\\":\\"focusable-no-name\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element is in tab order and does not have accessible text\\"}],\\"impact\\":\\"serious\\",\\"html\\":\\"<a href=\\\\\\"#\\\\\\"></a>\\",\\"target\\":[\\"a\\"],\\"failureSummary\\":\\"Fix all of the following:\\\\n  Element is in tab order and does not have accessible text\\\\n\\\\nFix any of the following:\\\\n  Element does not have text that is visible to screen readers\\\\n  aria-label attribute does not exist or is empty\\\\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\\\n  Element has no title attribute\\"}]}]"`;

exports[`a11y Results Formatter should format using specified options: 1 1`] = `
"\\"* (document-title) Documents must have <title> element to aid in navigation: html\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

\\"* (link-name) Links must have discernible text: a\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should format using specified options: 2 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should format using specified options: 3 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should format using specified options: 4 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 1`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 2`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;

exports[`a11y Results Formatter should throw error when violations are present 1`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/document-title 
	- WCAG Criteria: SA11Y-WCAG-SC2.4.2-P2

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.8/link-name 
	- WCAG Criteria: SA11Y-WCAG-SC4.1.2-P1"
`;
